File: src/network/protocol.py
from typing import Optional, List, Literal
from pydantic import BaseModel, Field

class HelloMessage(BaseModel):
    type: Literal["hello"] = "hello"
    protocol: int = 1
    client: str = "android"
    appVersion: str = "1.0.0"

class HelloAckMessage(BaseModel):
    type: Literal["hello_ack"] = "hello_ack"
    protocol: int = 1
    robotId: str
    robotName: str
    state: str

class DriveMessage(BaseModel):
    type: Literal["drive"] = "drive"
    direction: Literal["forward", "backward", "left", "right"]
    speed: float = Field(ge=-1.0, le=1.0)
    seq: Optional[int] = None

class TankDriveMessage(BaseModel):
    type: Literal["tank_drive"] = "tank_drive"
    left: float = Field(ge=-1.0, le=1.0)
    right: float = Field(ge=-1.0, le=1.0)
    seq: Optional[int] = None

class StopMessage(BaseModel):
    type: Literal["stop"] = "stop"
    seq: Optional[int] = None

class EmergencyStopMessage(BaseModel):
    type: Literal["emergency_stop"] = "emergency_stop"

class EmergencyResetMessage(BaseModel):
    type: Literal["emergency_reset"] = "emergency_reset"

class MotorTestMessage(BaseModel):
    type: Literal["motor_test"] = "motor_test"
    motor: int = Field(ge=1, le=4)
    speed: float = Field(ge=-1.0, le=1.0)

class HeartbeatMessage(BaseModel):
    type: Literal["heartbeat"] = "heartbeat"
    timestamp: Optional[float] = None

class HeartbeatAckMessage(BaseModel):
    type: Literal["heartbeat_ack"] = "heartbeat_ack"
    timestamp: float

class CommandAckMessage(BaseModel):
    type: Literal["command_ack"] = "command_ack"
    seq: Optional[int] = None
    accepted: bool
    reason: Optional[str] = None

class SafetyEventMessage(BaseModel):
    type: Literal["safety_event"] = "safety_event"
    event: Literal["person_detected", "person_clear", "emergency_stop", "camera_error", "connection_lost"]
    confidence: Optional[float] = None

class TelemetryMessage(BaseModel):
    type: Literal["telemetry"] = "telemetry"
    robotState: str
    safetyState: str
    speed: float
    leftSpeed: float
    rightSpeed: float
    m1: float
    m2: float
    m3: float
    m4: float
    personDetected: bool
    personConfidence: float
    cameraOk: bool
    detectorOk: bool
    oledOk: bool
    motorOk: bool
    cameraFps: float
    inferenceFps: float
    cpuTemp: Optional[float] = None
    cpuUsage: Optional[float] = None
    memoryUsage: Optional[float] = None
    wifiSignal: Optional[int] = None
    uptime: float = 0.0
    timestamp: float = 0.0
    personBbox: Optional[List[float]] = None

class ErrorMessage(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str

class PairRequest(BaseModel):
    pairCode: str
    clientInfo: Optional[str] = "Android Device"

class PairResponse(BaseModel):
    success: bool
    token: Optional[str] = None
    robotId: str
    robotName: str
    message: Optional[str] = None

class HealthResponse(BaseModel):
    status: str
    robotState: str
    safetyState: str
    camera: bool
    detector: bool
    motor: bool
    oled: bool
    server: bool

class InfoResponse(BaseModel):
    robotId: str
    robotName: str
    apiVersion: str
    protocolVersion: int
    hostname: str
    serverVersion: str
    paired: bool

File: android-app/src/types/protocol.ts
export type RobotStateType =
  | 'BOOTING'
  | 'READY'
  | 'PAIRING'
  | 'CONNECTED'
  | 'FORWARD'
  | 'BACKWARD'
  | 'TURN_LEFT'
  | 'TURN_RIGHT'
  | 'STOPPED'
  | 'PERSON_DETECTED'
  | 'SAFETY_STOP'
  | 'CAMERA_ERROR'
  | 'MOTOR_ERROR'
  | 'SYSTEM_ERROR'
  | 'DISCONNECTED'
  | 'SHUTTING_DOWN';

export type SafetyStateType = 'CLEAR' | 'PERSON_DETECTED' | 'SAFETY_STOP' | 'ERROR_STOP';

export interface TelemetryData {
  type: 'telemetry';
  robotState: RobotStateType;
  safetyState: SafetyStateType;
  speed: number;
  leftSpeed: number;
  rightSpeed: number;
  m1: number;
  m2: number;
  m3: number;
  m4: number;
  personDetected: boolean;
  personConfidence: number;
  cameraOk: boolean;
  detectorOk: boolean;
  oledOk: boolean;
  motorOk: boolean;
  cameraFps: number;
  inferenceFps: number;
  cpuTemp?: number | null;
  cpuUsage?: number | null;
  memoryUsage?: number | null;
  wifiSignal?: number | null;
  uptime: number;
  timestamp: number;
  personBbox?: [number, number, number, number] | null;
}

export interface SafetyEventData {
  type: 'safety_event';
  event: 'person_detected' | 'person_clear' | 'emergency_stop' | 'camera_error' | 'connection_lost';
  confidence?: number;
}

export interface CommandAckData {
  type: 'command_ack';
  seq?: number;
  accepted: boolean;
  reason?: string;
}

export interface HelloAckData {
  type: 'hello_ack';
  protocol: number;
  robotId: string;
  robotName: string;
  state: RobotStateType;
}

export interface ErrorData {
  type: 'error';
  code: string;
  message: string;
}

export interface PairResponse {
  success: boolean;
  token?: string;
  robotId: string;
  robotName: string;
  message?: string;
}

export interface HealthResponse {
  status: string;
  robotState: RobotStateType;
  safetyState: SafetyStateType;
  camera: boolean;
  detector: boolean;
  motor: boolean;
  oled: boolean;
  server: boolean;
}

export interface InfoResponse {
  robotId: string;
  robotName: string;
  apiVersion: string;
  protocolVersion: number;
  hostname: string;
  serverVersion: string;
  paired: boolean;
}
